File: lintcode/word-search.py
#!/usr/bin/env python
# encoding: utf-8

# Question: http://www.lintcode.com/en/problem/word-search/

import copy

class Solution:
    def dfs(self, board, word, path, x, y, i):
        if (x, y) in path:
            return False

        if i + 1 == len(word):
            return True

        path.add((x, y))
        if x + 1 >= 0 and x + 1 < len(board) and board[x + 1][y] == word[i + 1] and self.dfs(board, word, path, x + 1, y, i + 1):
            return True
        if x - 1 >= 0 and x - 1 < len(board) and board[x - 1][y] == word[i + 1] and self.dfs(board, word, path, x - 1, y, i + 1):
            return True
        if y + 1 >= 0 and y + 1 < len(board[0]) and board[x][y + 1] == word[i + 1] and self.dfs(board, word, path, x, y + 1, i + 1):
            return True
        if y - 1 >= 0 and y - 1 < len(board[0]) and board[x][y - 1] == word[i + 1] and self.dfs(board, word, path, x, y - 1, i + 1):
            return True
        path.remove((x, y))

        return False

    # @param board, a list of lists of 1 length string
    # @param word, a string
    # @return a boolean
    def exist(self, board, word):
        # write your code here
        if len(board) == 0 or len(board[0]) == 0:
            return False

        for i in xrange(len(board)):
            for j in xrange(len(board[0])):
                if board[i][j] == word[0]:
                    path = set()
                    if self.dfs(board, word, path, i, j, 0):
                        return True

        return False

if __name__ == '__main__':
    s = Solution()
    print s.exist(["ABCE","SFCS","ADEE"], "ABCB")
    print s.exist(["ABCE","SFCS","ADEE"], "ABCCED")

File: leetcode-2/word-search.cc
// Question: https://leetcode.com/problems/word-search/

class Solution {
public:
    bool exist(vector<vector<char>>& board, string word) {
        if (board.size() == 0 || board[0].size() == 0) {
            return false;
        }

        for (int i = 0; i < board.size(); ++i) {
            for (int j = 0; j < board[0].size(); ++j) {
                if (helper(board, word, 0, i, j)) {
                    return true;
                }
            }
        }
        return false;
    }

    bool helper(vector<vector<char>>& board, string& word, int idx, int i, int j) {
        if (idx == word.size()) {
            return true;
        }

        if (i < 0 || j < 0 || i >= board.size() || j >= board[0].size() || idx > (int)word.size()-1 || word[idx] != board[i][j]) {
            return false;
        }

        board[i][j] = '*';
        vector<int> row_delta{0,1,0,-1};
        vector<int> col_delta{1,0,-1,0};
        for (int k = 0; k < 4; ++k) {
            if (helper(board, word, idx+1, i+row_delta[k], j+col_delta[k])) {
                return true;
            }
        }
        board[i][j] = word[idx];
        return false;
    }
};
